AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift

```diff
@@ -600,42 +609,55 @@ extension AWSMobileClient {
+    internal func performHostedUISuccessfulSignInTasks(disableFederation: Bool = false,
+                                                       session: AWSCognitoAuthUserSession,
+                                                       federationToken: String,
+                                                       federationProviderIdentifier: String? = nil,
+                                                       signInInfo: inout [String: String]) -> AWSTask<AWSCredentials>  {
         federationDisabled = disableFederation
         if federationProviderIdentifier == nil {
             self.cachedLoginsMap = [self.userPoolClient!.identityProviderName: federationToken]
         } else {
             self.cachedLoginsMap = [federationProviderIdentifier!: federationToken]
         }
-        self.mobileClientStatusChanged(userState: .signedIn, additionalInfo: signInInfo)
-        self.saveLoginsMapInKeychain()
-        self.setLoginProviderMetadataAndSaveInKeychain(provider: .hostedUI)
-        self.internalCredentialsProvider?.clearCredentials()
-        self.internalCredentialsProvider?.credentials()
+        return postSignInKeychainAndCredentialsUpdate(provider: .hostedUI, additionalInfo:signInInfo)
     }
     
-    internal func performFederatedSignInTasks(provider: String, token: String) {
+    internal func performFederatedSignInTasks(provider: String, token: String) -> AWSTask<AWSCredentials>  {
         self.cachedLoginsMap = [provider:token]
-        self.mobileClientStatusChanged(userState: .signedIn, additionalInfo: [ProviderKey:provider, TokenKey: token])
         self.federationProvider = .oidcFederation
+        return postSignInKeychainAndCredentialsUpdate(provider: .oidcFederation, additionalInfo:[ProviderKey:provider, TokenKey: token])
+    }
+    
+    /// Post signin operations. Saves the login maps and provider metadata in keychain. Updates the credentials.
+    /// If credential update is successfull, informs the listener with UserState.signedIn status.
+    ///
+    /// - Parameters
+    ///     - provider: provider to be updated in keychain.
+    ///     - additionalInfo: additional infomap to be passed in signedIn callback.
+    func postSignInKeychainAndCredentialsUpdate(provider: FederationProvider,
+                                                additionalInfo: [String: String]) -> AWSTask<AWSCredentials> {
         self.saveLoginsMapInKeychain()
-        self.setLoginProviderMetadataAndSaveInKeychain(provider: .oidcFederation)
+        self.setLoginProviderMetadataAndSaveInKeychain(provider: provider)
         self.internalCredentialsProvider?.clearCredentials()
-        self.internalCredentialsProvider?.credentials()
+        return self.internalCredentialsProvider!.credentials().continueWith(block: { (task) -> AWSTask<AWSCredentials>? in
+            if task.result != nil {
```

Comment:
1. Should we invoke `mobileClientStatusChanged` if the result is `nil`? Is there any meaningful assertion we can make at that point?
1. Does it make sense for this level of the code to know anything about `mobileClientStatusChanged` in the first place?

Reply:
Moved the `mobileClientStatusChanged` to the place where completion handler is called.